rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md

```diff
@@ -0,0 +1,137 @@
+
+Relying on an external library will reduce the amount of code and complexity that
+currently exists within Vector and allow us to focus less on making to tools correct
+and more on ensuring we're using them properly.
+
+### Existing crate vs our implementation
+
+Leveraging `kube` was considered at the start of the Kubernetes integration project,
+but eventually we wrote our own implementation on top of the `k8s-openapi` crate.
+
+Over the past year and a half, `kube` has matured greatly and today it is being
+[donated](https://github.com/kube-rs/kube-rs/issues/584) to the [CNCF](https://www.cncf.io/).
+This appears to be the [state of `kube` roughly at the time](https://github.com/kube-rs/kube-rs/tree/c38d82162d2626bcfcf2ef8cf0c9d93e0734af49)
+of writing our own implementation. While our existing implementation is quite
+generic and modular, the needs of the component today are quite limited.
+Realistically we just need the following:
+
+- Client: authentication and configuration to call the Kubernetes API
+  - `kube-client`
+- Reflector: error handling and a persistent cache for an event stream
+  - `kube-runtime::reflector`
+
+## Drawbacks
+
+Converting the underlying libraries to `kube` (where possible) doesn't guarantee
+a resolution for reported bugs, but it does shift a non-trivial amount of functionality
+out of our project and onto a specialized library.
+
+## Prior Art
+
+- [Vector `kubernetes_logs` source](https://github.com/vectordotdev/vector/tree/master/src/sources/kubernetes_logs)
+  - [Vector `kubernetes`](https://github.com/vectordotdev/vector/tree/master/src/kubernetes)
+- [Datadog Kubelet provider](https://github.com/DataDog/datadog-agent/blob/main/pkg/autodiscovery/providers/kubelet.go)
+- [Fluent-Bit Kubernetes filter](https://github.com/fluent/fluent-bit/tree/master/plugins/filter_kubernetes)
+
+## Alternatives
+
+### Not do this at all
+
+Troubleshooting and maintaining the existing source has been challenging, and there is
+a clear lack of understanding of the current implementation. The source is an important
+part of our OSS offering and frequently used with the Agent role. Not doing this impairs
+our ability to support Vector running as an Agent in Kubernetes and reflects poorly on
+our overall reputation.
+
+### Rewrite to integrate with the Kubelet API
+
+This avenue was recommended internally, but is a departure from our existing implementation,
+and thus contains more unknowns and increases the likelihood of breaking changes to the
+existing source.
+
+### Replace existing source with an enrichment transform
+
+This option is capable of being used regardless of what role Vector has been deployed
+in, but it is a larger change to implement and a larger update for users too. Longer
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to re-evaluate the usage of `evmap` in this component?
```

Comment:
Do you have an answer for this?

Reply:
I _don't_ think I do, I don't remember seeing any of the reasoning why we picked `evmap` originally. It seems like it's possibly higher performance depending on usage. We could probably benchmark further down the line, `kubernetes_logs` is the only thing in Vector that uses it, so it'd mostly be reducing dependencies.

**TL;DR:** probably fine keeping it as is, unless we strongly want to remove deps that may be superfluous.